src/cli/telemetry/telemetry.ts

```diff
@@ -137,9 +147,77 @@ export const prepareData = async (
     typescript,
     rollup,
     has_app_pwa_config,
+    config: anonymizedConfig,
   };
 };
 
+// Setting a key type to `never` excludes it from a mapped type, so we
+// can get only keys which map to a string value by excluding all keys `K`
+// where `d.Config[K]` does not extend `string`.
+type ConfigStringKeys = keyof {
+  [K in keyof d.Config as Required<d.Config>[K] extends string ? K : never]: d.Config[K];
+};
+
+/**
+ * Anonymize the config for telemtry, replacing potentially revealing config props
+ * with a placeholder string if they are present (this lets us still track how frequently
+ * these config options are being used)
+ *
+ * @param config the config to anonymize
+ * @returns an anonymized copy of the same config
+ */
+export const anonymizeConfigForTelemtry = (config: d.Config): d.Config => {
+  const anonymizedConfig = { ...config };
+  const propsToAnonymize: ConfigStringKeys[] = [
+    'rootDir',
+    'fsNamespace',
+    'packageJsonFilePath',
+    'namespace',
+    'srcDir',
+    'srcIndexHtml',
+    'buildLogFilePath',
+    'cacheDir',
+    'configPath',
+    'tsconfig',
+  ];
+
+  for (const prop of propsToAnonymize) {
+    if (anonymizedConfig[prop] !== undefined) {
+      anonymizedConfig[prop] = 'omitted';
+    }
+  }
+
+  anonymizedConfig.outputTargets = (config.outputTargets ?? []).map((target) => ({
```

Comment:
For anonymizing our output targets, the dev server, and the TypeScript compiler options, is there a way we could recursively apply something like `ConfigStringKeys` to anything that's an object literal? I'm a bit concerned about maintaining configurations where we would have to remember the anonymize certain items as we add/change these things.

Alternatively, we could always punt on that for now. ATM, we're really only concerned with the following fields for the purposes of determining which v3 flags to set (potentially) new defaults for:
- OutputTargetHydrate#external ([source](https://github.com/ionic-team/stencil/blob/04fb83090c968010a400cee96b129f067c5cb6f0/src/declarations/stencil-public-compiler.ts#L1867))
- OutputTargetDistCustomElements#autoDefineCustomElements ([source](https://github.com/ionic-team/stencil/blob/04fb83090c968010a400cee96b129f067c5cb6f0/src/declarations/stencil-public-compiler.ts#L1932))
- OutputTargetDistCustomElements#generateTypeDeclarations ([source](https://github.com/ionic-team/stencil/blob/04fb83090c968010a400cee96b129f067c5cb6f0/src/declarations/stencil-public-compiler.ts#L1936))

We _could_ temporarily hard code these as a temporary solution (if no recursive one is possible). In that scenario, I'd be OK with omitting the ts config and dev server config entirely for now.

Reply:
Created `STENCIL-469` for this

Reply:
I don't know if that covers all of our cases here 🤔 More immediately, we also need to consider output target fields since we're going to anonymize/serialize/deserialize those.

ATM, our output targets extend one of two base types, [`OutputTargetBase`](https://github.com/ionic-team/stencil/blob/c0401477c0074f280543bc19d7893a03ed1d5d91/src/declarations/stencil-public-compiler.ts#L2009) or [`OutputTargetBaseNext`
](https://github.com/ionic-team/stencil/blob/c0401477c0074f280543bc19d7893a03ed1d5d91/src/declarations/stencil-public-compiler.ts#L1971), and each output target can have multiple fields that are specific to the output target itself.  

Because of the number of fields, I think we may need to walk the entire object. Otherwise, I don't know if we can cover/anonymize every string field in the output targets.

I think this would look something like:
```typescript
const omitOutputTargetDetails = (outputTargetPiece: any): any => {
  if (!Array.isArray(outputTargetPiece) && typeof outputTargetPiece === 'object') {
    for (const [key, value] of Object.entries(outputTargetPiece)) {
      if (typeof value === 'string') {
        outputTargetPiece[key] = 'omitted';
     } else if (Array.isArray(value)) {
        outputTargetPiece[key] = 'omitted';
    } else if (typeof outputTargetPiece === 'object') {
        outputTargetPiece[key] = omitOutputTargetDetails(outputTargetPiece[key]);
    }
  }
}; 
``` 

and call it like so:

```typescript
anonymizedConfig.outputTargets = (config.outputTargets ?? []).map((outputTarget) => omitOutputTargetDetails(outputTarget));
```

or something like that (this was all back of the napkin prototyping).  But LMK if there's something I'm missing here (that's very possible)

Reply:
what if instead of trying to omit all of the fields on the output targets that could possibly hold identifying information, we instead opt-in to the specific fields we're interested in looking at? Something like

```ts
  anonymizedConfig.outputTargets = (config.outputTargets ?? []).map((target) => {
    const { external, autoDefineCustomElements, generateTypeDeclarations } = target;
    return { external, autoDefineCustomElements, generateTypeDeclarations };
});
```

might have to do something a bit more complicated to deal with the types, but that would be the basic idea

Reply:
maybe we'd want `target.type` in there too

Reply:
Agreed, we'd want to keep `target.type` in there for sure.  My vote here is for us to try to be an encompassing as possible for the output targets now. If we can find/use a generalized solution (that's not too rough on performance) we should pursue that route.  Ideally I'd like us to not have to add a new key to telemetry code when we modify the shape of output targets. The reason being that in addition to having to remember/decide if the field is important enough to serialize, I think we can run into some issues with typings where a field like `autoDefineCustomElements` doesn't exist on every output target.

That said, that solution isn't all too perfect either. If we go the other way and want to omit a field entirely (like the dev server), we then _do_ have to remember to strike it from the telemetry list

Reply:
ok so a little investigation! I implemented two different ways to recursively omit any keys in the output target objects which aren't on a 'keep' list that we define:

```ts
const OUTPUT_TARGET_KEYS_TO_KEEP = ['type', 'typesDir', 'baseUrl'];
```

The first is based on what you suggested:

```ts
  const omitOutputTargetDetails = (outputTargetPiece: any): any => {
    if (!Array.isArray(outputTargetPiece) && typeof outputTargetPiece === 'object') {
      for (const [key, value] of Object.entries(outputTargetPiece)) {
        if (typeof value === 'string') {
          if (!OUTPUT_TARGET_KEYS_TO_KEEP.includes(key)) {
            outputTargetPiece[key] = 'omitted';
          }
        } else if (Array.isArray(value)) {
          outputTargetPiece[key] = 'omitted';
        } else if (typeof outputTargetPiece === 'object') {
          outputTargetPiece[key] = omitOutputTargetDetails(outputTargetPiece[key]);
        }
      }
    }
    return outputTargetPiece;
  };

  anonymizedConfig.outputTargets = (config.outputTargets ?? []).map(omitOutputTargetDetails);
```

the second uses `JSON.stringify`, passing the optional [2nd argument](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#the_replacer_parameter):


```ts
  anonymizedConfig.outputTargets = JSON.parse(
    JSON.stringify(config.outputTargets ?? [], (key, value) => {
      if (!['string', 'number'].includes(typeof value)) {
        return value;
      }
      if (OUTPUT_TARGET_KEYS_TO_KEEP.includes(key)) {
        return value;
      }
      return 'omitted';
    })
  );
```

I wrote a little hacky 'performance' test just to get a ballpark idea of how these perform:

```ts
  it.only('test performance of OT anonymizers', () => {
    const config = {
      flags: {
        args: [],
      },
      outputTargets: [
        { type: 'www', baseUrl: 'https://example.com', serviceWorker: { swDest: './tmp' } }
      ],
    } as d.Config;    

    let beforeJSON = new Date();
    for (let i = 0; i < 1000; i++) {
      anonymizeConfigForTelemetryJSON(config)
    }
    let afterJSON = new Date()
    // @ts-ignore
    let JSONDifference = afterJSON - beforeJSON;
    console.log(`JSON version took ${JSONDifference}`);

    let beforenonJSON = new Date();
    for (let i = 0; i < 10000; i++) {
      anonymizeConfigForTelemetry(config)
    }
    let afternonJSON = new Date()
    // @ts-ignore
    let nonJSONDifference = afternonJSON - beforenonJSON;
    console.log(`nonJSON version took ${nonJSONDifference}`);
  });
```

The results were kind of surprising, but possibly make sense? I thought that the `JSON.parse(JSON.stringify( ... ))` would be slower because it's basically doing some 'unnecessary' overhead work serializing and deserializing, but actually that version is quite a bit faster - anonymizing the above config 10000 times takes ~4-5ms with that version vs ~31-35ms for the recursive function call version. I suspect this is just because w/ `JSON.stringify` the descent through the object is handled by native code, rather than recursive function calls, `Object.entries`, etc.

So anywhoo - that's what I've found. I _personally_ find the `JSON.{stringify,parse}` version to be quite a bit more elegant, letting us just write a little function which decides what to do based on each `k, v` pair (sort of a context-free way to handle it, if you will) and we don't have to include any logic for iterating, recurring, etc.

What do you think?